### PR TITLE
Single functions to get/set every celestial vault element

### LINKED ISSUE #11917: Utilities to return and set every celestial vault settings in one function
## Problem
If I want to get every celestial vault element of a player, to know the status of each element, right now I have to do
```lua
local p_cvault = {}

    -- TODO: Minetest ?.?: get_sky() doesn't work properly, instead of being
    -- able to do sky = player:get_sky(), I need to do all this.
    -- Relevant issue: https://github.com/minetest/minetest/issues/11890
    p_cvault.sky = {}
    p_cvault.base_color = (type(get_sky[1]) ~= "table") and get_sky[1] or table.copy(get_sky[1])
    p_cvault.type = get_sky[2]
    p_cvault.sky_color = player:get_sky_color()
    p_cvault.textures = table.copy(get_sky[3])
    p_cvault.clouds = get_sky[4]

    p_cvault.sun = player:get_sun()
    p_cvault.moon = player:get_moon()
    p_cvault.stars = player:get_stars()
    p_cvault.clouds = player:get_clouds()
```

## Solutions
Add a `player:get_celestialvault()` function returning a table containing all the different tables returned by the other functions such as  
`{sky = {...}, sun = {...}, moon = {...}, stars = {...}, clouds = {...}}`  

This could be done either by adding a Lua utility function in the builtin, or by hardcoding it in C++. Core devs choice (if they agree with the whole idea, of course).  

This can also apply to a `player:set_celestialvault()` function, requiring the same table structure.

### Usecase
Usecase: store all the celestial vault elements of a player in a temporary table to later re-apply them (i.e. when entering a minigame)

### PR DESCRIPTION
**Pull request by Zughy**
- Goal of the PR: have a way to instantly retrieve/set every celestial vault element in one shot
- How does the PR work? implements `get_celestial_vault()` and `set_celestial_vault()`
- Does it resolve any reported issue? Fixes #11917
- Does this relate to a goal in [the roadmap](../doc/direction.md)? 2.4, API convenience

I've used private functions for the set part, as I was having issues with `readParam`, but if there is a problem of consistence (the rest is `static void`), just tell me and I'll try to make all the functions of the same nature

Another name could be `get/set_skybox()`, but in that case the internal `SkyboxParams` struct should be renamed to `SkyParams` to avoid misunderstandings

## To do

This PR is Ready for Review.

## How to test
```lua
minetest.register_on_joinplayer(function(player)
    minetest.chat_send_all("Day")
    minetest.set_timeofday(0.35)
    player:set_sky({sky_color = {day_sky = "#eeeeee"}})
    --player:set_sky("#eeeeee", "plain")    -- if you wanna test retrocompatibility (color doesn't work, not this PR fault)
    player:set_sun({scale=2})
    player:set_clouds({color="#cccceecc"})
    
    minetest.after(2, function()
        minetest.chat_send_all("Vault has changed")             -- sky and clouds aren't reset as they're not specified
        player:set_celestial_vault({
            sun = {scale=2.5}
        })
            
        minetest.after(2, function()
            minetest.chat_send_all("Night")
            minetest.set_timeofday(0.85)
            player:set_moon({scale=2})
            player:set_stars({count = 3000})
            
            minetest.after(2, function()
                minetest.chat_send_all("Vault has changed")     -- stars aren't reset as they're not specified
                player:set_celestial_vault({
                    moon = {scale=1.2},
                    clouds = {denity=0.75}
                })
                
                minetest.after(2, function()
                    minetest.chat_send_all(dump(player:get_celestial_vault()) .. "\n\n") --instead of doing...
                    --[[minetest.chat_send_all(dump(player:get_sky()))
                    minetest.chat_send_all(dump(player:get_sun()))
                    minetest.chat_send_all(dump(player:get_moon()))
                    minetest.chat_send_all(dump(player:get_stars()))
                    minetest.chat_send_all(dump(player:get_clouds()))]]
                    player:set_celestial_vault() --instead of doing...
                    --[[player:set_sky()
                    player:set_sun()
                    player:set_moon()
                    player:set_stars()
                    player:set_clouds()]]
                    minetest.chat_send_all("Vault has been reset")
                end)
            end)
        end)
    end)
end)
```

### CODE REVIEW

File: src/script/lua_api/l_object.cpp [hunk 1]
```diff
@@ -1716,32 +1716,25 @@ int ObjectRef::l_hud_get_hotbar_selected_image(lua_State *L)
 	return 1;
 }
 
-// set_sky(self, sky_parameters)
-int ObjectRef::l_set_sky(lua_State *L)
+void ObjectRef::set_sky(lua_State *L, int table_index, RemotePlayer *player)
 {
-	NO_MAP_LOCK_REQUIRED;
```
Comment (appgurueu) [inline src/script/lua_api/l_object.cpp:1722]: Why does this suddenly require a map lock?

Comment (sfan5) [reply src/script/lua_api/l_object.cpp:1722]: it doesn't, these are stubs

they should be kept vaguely correct anyway

**Comment by SmallJoker** [comment]
Concept approved in the meeting.
https://irc.minetest.net/minetest-dev/2022-04-24#i_5965626

**Comment by Zughy** [comment]
Rebased

**Comment by Zughy** [comment]
I can rebase for the third time next week, but it'd be nice to have someone reviewing it at some point

**Comment by Zughy** [comment]
I'd really like to avoid rebasing this for the fourth time, anyone?

**Comment by Zughy** [comment]
A simpler approach would be to create a lua utility function calling `set_sky`, `set_sun` ecc, without touching any C++ code. Your call

**Comment by appgurueu** [comment]
> but I guess it'd be slower?

I wouldn't be so sure about this. LuaJIT may surprise you.

File: doc/lua_api.txt [hunk 2]
```diff
@@ -7259,6 +7259,18 @@ child will follow movement and rotation of that bone.
...
+    * `vault_parameters` is a table optionally containing the tables `sky`, `sun`,
+      `moon`, `stars` and `clouds`. These tables take in the parameters
+      indicated in their more specific functions (`set_sky()`, `set_sun()` etc).
+   * Passing no arguments resets every element of the vault to its default values.
+     To reset just a single element, its specific function must be used instead.
+* `get_celestial_vault()`
+    * returns a table containing every celestial vault element of the player, in
+      the format `sky`, `sun`, `moon`, `stars` and `clouds` as tables. See their
+      corrispondent specific functions to know what each one of them contains (`get_sky()`, 
+      `get_sun()` etc).
```
Comment (sfan5) [inline doc/lua_api.txt:7273]:
```suggestion
    * returns a table containing every celestial vault element of the player, with
      `sky`, `sun`, `moon`, `stars` and `clouds` as tables. See their correspondent
       functions to know what each one of them contains (`get_sky()`, `get_sun()` etc).
```

File: src/script/lua_api/l_object.cpp [hunk 3]
```diff
@@ -1705,32 +1705,25 @@ int ObjectRef::l_hud_get_hotbar_selected_image(lua_State *L)
...
 	SkyboxParams sky_params = player->getSkyParams();
 
 	// reset if empty
-	if (lua_isnoneornil(L, 2) && lua_isnone(L, 3)) {
+	if (lua_isnoneornil(L, table_index)) {
 		sky_params = SkyboxDefaults::getSkyDefaults();
-	} else if (lua_istable(L, 2) && !is_color_table(L, 2)) {
-		lua_getfield(L, 2, "base_color");
+	} else {
+		lua_getfield(L, -1, "base_color");
```
Comment (sfan5) [inline src/script/lua_api/l_object.cpp:1716]: if you add `table_index` then you should use it here

File: src/script/lua_api/l_object.cpp [hunk 4]
```diff
@@ -1798,10 +1791,312 @@ int ObjectRef::l_set_sky(lua_State *L)
...
+
+void ObjectRef::set_sun(lua_State *L, int table_index, RemotePlayer *player)
+{
+	SunParams sun_params = player->getSunParams();
+
+	// reset if empty
+	if (lua_isnoneornil(L, table_index)) {
+		sun_params = SkyboxDefaults::getSunDefaults();
+	} else {
+		luaL_checktype(L, -1, LUA_TTABLE);
```
Comment (sfan5) [inline src/script/lua_api/l_object.cpp:1807]: -1 -> `table_index` here and in the other functions too

File: src/script/lua_api/l_object.cpp [hunk 5]
```diff
@@ -1798,10 +1791,312 @@ int ObjectRef::l_set_sky(lua_State *L)
...
+{
+	NO_MAP_LOCK_REQUIRED;
+	ObjectRef *ref = checkObject<ObjectRef>(L, 1);
+	RemotePlayer *player = getplayer(ref);
+	if (player == nullptr)
+		return 0;
+
+	// reset the whole vault if empty
+	if (lua_isnoneornil(L, 2)) {
+		set_sky(L, 3, player);
```
Comment (sfan5) [inline src/script/lua_api/l_object.cpp:2025]: there's no third argument

File: src/script/lua_api/l_object.cpp [hunk 6]
```diff
@@ -1798,10 +1791,312 @@ int ObjectRef::l_set_sky(lua_State *L)
...
+	lua_setfield(L, -2, "thickness");
+	lua_newtable(L);
+	lua_pushnumber(L, cloud_params.speed.X);
+	lua_setfield(L, -2, "x");
+	lua_pushnumber(L, cloud_params.speed.Y);
+	lua_setfield(L, -2, "y");
+	lua_setfield(L, -2, "speed");
+}
+
+//set_celestial_vault(self)
```
Comment (sfan5) [inline src/script/lua_api/l_object.cpp:2014]:
```suggestion
// set_celestial_vault(self, parameter table)
```

**Review by sfan5 — CHANGES_REQUESTED** [review]
probably fine otherwise but didn't test

**Comment by Zughy** [comment]
I've lost interest in this PR after 3 rebases, the fact that it won't make it for 5.7 (so at least another month waiting), and the idea that I might face more change requests, prolonging the agony even more

**Pull request closed without merging**